src/algorithm/fixed_window.go

```diff
@@ -0,0 +1,128 @@
+			DurationUntilReset: utils.CalculateFixedReset(limit.Limit, fw.timeSource),
+		}
+	}
+
+	isOverLimit, limitRemaining, durationUntilReset := fw.IsOverLimit(limit, int64(results), hitsAddend)
+
+	if !isOverLimit {
+		return &pb.RateLimitResponse_DescriptorStatus{
+			Code:               pb.RateLimitResponse_OK,
+			CurrentLimit:       limit.Limit,
+			LimitRemaining:     uint32(limitRemaining),
+			DurationUntilReset: utils.CalculateFixedReset(limit.Limit, fw.timeSource),
+		}
+	} else {
+		if fw.localCache != nil {
+			durationUntilReset = utils.MaxInt(1, durationUntilReset)
+
+			err := fw.localCache.Set([]byte(key), []byte{}, durationUntilReset)
+			if err != nil {
+				logger.Errorf("Failing to set local cache key: %s", key)
+			}
+		}
+
+		return &pb.RateLimitResponse_DescriptorStatus{
+			Code:               pb.RateLimitResponse_OVER_LIMIT,
+			CurrentLimit:       limit.Limit,
+			LimitRemaining:     uint32(limitRemaining),
+			DurationUntilReset: utils.CalculateFixedReset(limit.Limit, fw.timeSource),
+		}
+	}
+}
+
+func (fw *FixedWindowImpl) IsOverLimit(limit *config.RateLimit, results int64, hitsAddend int64) (bool, int64, int) {
+	limitAfterIncrease := results
+	limitBeforeIncrease := limitAfterIncrease - int64(hitsAddend)
+	overLimitThreshold := int64(limit.Limit.RequestsPerUnit)
+	nearLimitThreshold := int64(math.Floor(float64(float32(overLimitThreshold) * fw.nearLimitRatio)))
+
+	if limitAfterIncrease > overLimitThreshold {
+		if limitBeforeIncrease >= overLimitThreshold {
+			fw.PopulateStats(limit, 0, uint64(hitsAddend), 0)
+		} else {
+			fw.PopulateStats(limit, uint64(overLimitThreshold-utils.MaxInt64(nearLimitThreshold, limitBeforeIncrease)), uint64(limitAfterIncrease-overLimitThreshold), 0)
+		}
+
+		return true, 0, int(utils.UnitToDivider(limit.Limit.Unit))
+	} else {
+		if limitAfterIncrease > nearLimitThreshold {
+			if limitBeforeIncrease >= nearLimitThreshold {
+				fw.PopulateStats(limit, uint64(hitsAddend), 0, 0)
+			} else {
+				fw.PopulateStats(limit, uint64(limitAfterIncrease-nearLimitThreshold), 0, 0)
+			}
+		}
+
+		return false, overLimitThreshold - limitAfterIncrease, int(utils.UnitToDivider(limit.Limit.Unit))
+	}
+}
+
+func (fw *FixedWindowImpl) IsOverLimitWithLocalCache(key string) bool {
+	if fw.localCache != nil {
+		_, err := fw.localCache.Get([]byte(key))
+		if err == nil {
+			return true
+		}
+	}
+	return false
+}
+
+func (fw *FixedWindowImpl) GenerateCacheKeys(request *pb.RateLimitRequest,
+	limits []*config.RateLimit, hitsAddend int64) []utils.CacheKey {
+	return fw.cacheKeyGenerator.GenerateCacheKeys(request, limits, uint32(hitsAddend), fw.timeSource.UnixNow())
+}
+
+func (fw *FixedWindowImpl) PopulateStats(limit *config.RateLimit, nearLimit uint64, overLimit uint64, overLimitWithLocalCache uint64) {
+	limit.Stats.NearLimit.Add(nearLimit)
+	limit.Stats.OverLimit.Add(overLimit)
+	limit.Stats.OverLimitWithLocalCache.Add(overLimitWithLocalCache)
+}
+
+func (fw *FixedWindowImpl) GetNewTat() int64 {
```

Comment:
yes, actually I have difficultly getting data from the rolling window since Redis/Memcached implementation for rolling window need information that fixed window doesn't need to update the data in the backend. will take a look at this one.